### ISSUE TITLE
Custom Adapters

### ISSUE THREAD

**Issue by olmohake** [better-auth/better-auth#785]
Our data model requires the entries in the user table to point to entries in the partner table via a foreign key reference. It is currently not possible to implement this, with the default db connectors.  It would thus be very helpful to enable users to build their own adapters to work around such limitations.

**Comment by angelhodar** [comment]
I was going to say something similar, in my case I am using MongoDB. Maybe some configuration functions to be able to abstract database interaction would be amazing

**Comment by ghost2023** [comment]
Hey, I was going through the codebase when I saw the comment. I am not a maintainer but I think you can build your own custom adapter that implement the Adapter in [adapter.ts](https://github.com/better-auth/better-auth/blob/main/packages/better-auth/src/types/adapter.ts) line 26. 

And also you can see the way other adapter were implemented too.


```typescript
export interface Adapter {
	id: string;
	create: <T extends Record<string, any>, R = T>(data: {
		model: string;
		data: T;
		select?: string[];
	}) => Promise<R>;
	findOne: <T>(data: {
		model: string;
		where: Where[];
		select?: string[];
	}) => Promise<T | null>;
	findMany: <T>(data: {
		model: string;
		where?: Where[];
		limit?: number;
		sortBy?: {
			field: string;
			direction: "asc" | "desc";
		};
		offset?: number;
	}) => Promise<T[]>;
	/**
	 * ⚠︎ Update may not return the updated data
	 * if multiple where clauses are provided
	 */
	update: <T>(data: {
		model: string;
		where: Where[];
		update: Record<string, any>;
	}) => Promise<T | null>;
	updateMany: (data: {
		model: string;
		where: Where[];
		update: Record<string, any>;
	}) => Promise<number>;
	delete: <T>(data: { model: string; where: Where[] }) => Promise<void>;
	deleteMany: (data: { model: string; where: Where[] }) => Promise<number>;
	/**
	 *
	 * @param options
	 * @param file - file path if provided by the user
	 * @returns
	 */
	createSchema?: (
		options: BetterAuthOptions,
		file?: string,
	) => Promise<{
		code: string;
		fileName: string;
		append?: boolean;
		overwrite?: boolean;
	}>;
	options?: Record<string, any>;
}
```

**Comment by octet-stream** [comment]
Yeah, this is doable - you can build your own adapter. I'm currently working on Mikro ORM adapter for Better Auth (I will share it once I finish the adapter and port it to Better Auth repo, and after it passes the test suite), however as far as I can tell, Better Auth doesn't support joins - it just calls adapter multiple times (one call per requested data).

**Comment by olmohake** [comment]
It’s doable. However one has reimplement all functions of an adapter at the moment, even if one wants t to support only one or two custom functions.
Exporting the adapters or a wrapper would be extremely helpful to avoid reimplementing the same functionality of the generic functions:

```ts
export const customAdapter = ({name,resolvers, db, options}) => {
    const adapter = generic_adapter(db, options) //initialize adapter according to db & option configurations
    return {
        id: name,
        async create({model, values, select}) {
            if(resolvers[model]?.create){
                return await resolvers[model].create({values, select})
            }
            return await adapter.create({model, values,select})
        }
    }
}
```

**Comment by matthewmorek** [comment]
I was thinking exactly the same thing: it's possible to create your own adapter, but then you have to reimplement the internals, because there's just no other way to satisfy the requirement. A simple, decently documented adapter API is all w.e need

Right now, it doesn't make sense to port my custom OrchidORM adapter from AuthJS, and it's the only obstacle for me to switch over.

**Comment by oskar-gmerek** [comment]
I have created adapter for SurrealDB, source code: https://github.com/oskar-gmerek/surreal-better-auth.git maybe that will help you in anyway.

Im not pretty sure if I understand your needs, but I think that is already possible with hooks: https://www.better-auth.com/docs/concepts/hooks

**Comment by olmohake** [comment]
@Bekacru and co. Any chance better auth will implement this in the near future?

**Comment by ping-maxwell** [comment]
Hello all, we now have a wrapper function helpful for creating adapters.
You can read more here:
https://www.better-auth.com/docs/guides/create-a-db-adapter

I'll close this for now.

**Comment by WendaoLee** [comment]
> Hello all, we now have a wrapper function helpful for creating adapters.大家好，我们现在有一个有助于创建适配器 (adapter) 的包装函数 (wrapper function)。 You can read more here:您可以在这里阅读更多内容： https://www.better-auth.com/docs/guides/create-a-db-adapter
> 
> I'll close this for now.我先暂时关闭这个。

Thanks for your work!

But for the `create` method mentioned in document,the create's `T` seems will occur some type problem by its type signature:
```typescript
(property) CustomAdapter.create: <T extends Record<string, any>>({ data, model, select, }: {
    model: string;
    data: T;
    select?: string[];
}) => Promise<T>
```
For some type-sensitive ORM like `prisma`,if you not specific the create type,the logic of database handler won't works fine like:

```typescript
        create: async ({ data, model, select }) => {
         // type error
          return await prisma.user.create({
            data:{
              ...data
            }
          })
        },
```

If I code like this:

```typescript
        create: async ({ data, model, select }:{
          data:{
            id:string,
            name:string,
            email:string,
            emailVerified:boolean,
            image:string,
            createdAt:Date,
            updatedAt:Date
          },
          model:string,
          select?:string[]
        }) => {
          return await prisma.user.create({
            data:{
              ...data
            }
          })
        },

```

Also get type error:

![Image](https://github.com/user-attachments/assets/6d7f19e3-fc1e-422b-bf82-cbbee52f1313)

Is there a way to solve this?As the logic of database should write here.

**Comment by ping-maxwell** [comment]
@WendaoLee As long as you're sure you've implemented everything correctly, you can force-cast a type `any` on the return-type.

**Comment by WendaoLee** [comment]
> [@WendaoLee](https://github.com/WendaoLee) As long as you're sure you've implemented everything correctly, you can force-cast a type `any` on the return-type.

Is there any plan to optimize this in future?👀

**Comment by ping-maxwell** [comment]
Probably not, it's intentional. It's most ideal that the output type from an ORM to match what we expect. However if an ORM (such as yours) can't do that, then it's fine to force-cast.
